### PR TITLE
Update to go 1.21

### PR DESCRIPTION
**Pull request by Neo2308**
**What this PR does**:
Updated build-image/Dockerfile as instructed at https://cortexmetrics.io/docs/contributing/how-to-update-the-build-image/

**Which issue(s) this PR fixes**:

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by Neo2308** [comment]
Waiting for a maintainer to push the build image to quay (Steps 2-4 on https://cortexmetrics.io/docs/contributing/how-to-update-the-build-image/). Once that is done, I can make the rest of the changes in this PR.

File: build-image/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM golang:1.20.4-bullseye
+FROM golang:1.21-bullseye
```
Comment (friedrichg) [inline build-image/Dockerfile:1]:
```suggestion
FROM golang:1.21.0-bullseye
```

**Review by yeya24 — COMMENTED** [review]
Should we update all images we use to 1.21? Now I can still see we are using go 1.20 in CI

**Comment by yeya24** [comment]
Hi @Neo2308, sorry for the late response.

I have built the image based on this branch and pushed the image to quay.

This is the image. `quay.io/cortexproject/build-image:feature-master-update-to-go-1.21-050c36e7b`.

Please let me know if you are still willing to continue this. Thanks!

**Comment by yeya24** [comment]
Hey @Neo2308, I am sorry I will close this pr due to another pr for the same purpose https://github.com/cortexproject/cortex/pull/5618 got merged.

Thanks so much for your contribution and I apologize for any inconvenience I caused for not helping unblock this change timely.

**Pull request closed without merging**